src/components/date-time-input/date-time-input.ts

```diff
@@ -0,0 +1,652 @@
+        this.hasTimeParts,
+        this.hasDateParts
+      );
+
+      if (Object.keys(errors).length > 0) {
+        this.invalid = true;
+      } else {
+        this.invalid = false;
+      }
+    }
+
+    return errors;
+  }
+
+  private get hasDateParts(): boolean {
+    const parts =
+      this._inputDateParts ||
+      DateTimeUtil.parseDateTimeFormat(this.inputFormat);
+
+    return parts.some(
+      (p) =>
+        p.type === DateParts.Date ||
+        p.type === DateParts.Month ||
+        p.type === DateParts.Year
+    );
+  }
+
+  private get hasTimeParts(): boolean {
+    const parts =
+      this._inputDateParts ||
+      DateTimeUtil.parseDateTimeFormat(this.inputFormat);
+    return parts.some(
+      (p) =>
+        p.type === DateParts.Hours ||
+        p.type === DateParts.Minutes ||
+        p.type === DateParts.Seconds
+    );
+  }
+
+  private get targetDatePart(): DatePart | undefined {
+    let result;
+
+    if (this.hasFocus) {
+      const partType = this._inputDateParts.find(
+        (p) =>
+          p.start <= this.inputSelection.start &&
+          this.inputSelection.start <= p.end &&
+          p.type !== DateParts.Literal
+      )?.type as string as DatePart;
+
+      if (partType) {
+        result = partType;
+      }
+    } else {
+      if (this._inputDateParts.some((p) => p.type === DateParts.Date)) {
+        result = DatePart.Date;
+      } else if (this._inputDateParts.some((p) => p.type === DateParts.Hours)) {
+        result = DatePart.Hours;
+      } else {
+        result = this._inputDateParts[0].type as string as DatePart;
+      }
+    }
+
+    return result;
+  }
+
+  private get datePartDeltas(): DatePartDeltas {
+    return Object.assign({}, this._datePartDeltas, this.spinDelta);
+  }
+
+  constructor() {
+    super();
+    this.addEventListener('wheel', this.onWheel);
+  }
+
+  public override connectedCallback() {
+    super.connectedCallback();
+    this.updateDefaultMask();
+    this.setMask(this.inputFormat);
+    if (this.value) {
+      this.updateMask();
+    }
+  }
+
+  /** Checks for validity of the control and shows the browser message if it's invalid. */
+  public reportValidity() {
+    const state = this._value
+      ? Object.keys(this.validate()!).length === 0
+      : this.input.reportValidity();
+
+    this.invalid = !state;
+    return state;
+  }
+
+  /** Check for validity of the control */
+  public checkValidity() {
+    if (this.disabled) {
+      return this.input.checkValidity();
+    }
+
+    if (!this._value) {
+      return !this.required;
+    }
+
+    return (
+      this.input.checkValidity() && Object.keys(this.validate()!).length === 0
+    );
+  }
+
+  public stepUp(datePart?: DatePart, delta?: number): void {
+    const targetPart = datePart || this.targetDatePart;
+
+    if (!targetPart || this.readonly) {
+      return;
+    }
+
+    const newValue = this.trySpinValue(targetPart, delta);
+    this.value = newValue;
+    this.setFocus();
```

Comment:
The `setFocus` has nothing to do with focus and more with `setSelectionRange` 😃 
I suppose these methods are used internally and that is why they do more than they should (keeping/modifying state).

IMO you should change them to do only one thing and move the UI state/interactions in protected handlers.